parsing(c_parser): added support for major data types and shorthand operators

smitgajjar opened the pull request:
<!-- Your title above should be a short description of what
was changed. Do not include the issue number in the title. -->

#### References to other Issues or PRs
<!-- If this pull request fixes an issue, write "Fixes #NNNN" in that exact
format, e.g. "Fixes #1234" (see
https://tinyurl.com/auto-closing for more information). Also, please
write a comment on that issue linking back to this pull request once it is
open. -->


#### Brief description of what is fixed or changed
- Added support for data types in C parser:
    - `int` data types:
        - `signed char`(8-bit integer)
        - `unsigned char`(8-bit integer, unsigned)
        - `short` (16-bit integer)
        - `short int` (16-bit integer)
        - `signed short` (16-bit integer)
        - `signed short int` (16-bit integer)
        - `unsigned short` (16-bit integer, unsigned)
        - `unsigned short int` (16-bit integer, unsigned)
        - `unsigned int` (32-bit integer, unsigned)
        - `long` (64-bit integer)
        - `long int` (64-bit integer)
        - `signed long` (64-bit integer)
        - `signed long int` (64-bit integer)
        - `unsigned long` (64-bit integer, unsigned)
        - `unsigned long int` (64-bit integer, unsigned)

    - `float` data types:
        - `double` (double precision floating point)
        - `long double` (extended precision floating point)
- Added support for shorthand operators in C parser:
    - `+=`, `-=`, `*=`, `/=` and `%=`
 - Removed the assumption of value of a variable, if it is not initialized while variable declaration. Also, removed the assumption of default value of formal parameters while parameter declaration in function definition or function prototype

#### Other comments

#### Release Notes

<!-- Write the release notes for this release below. See
https://github.com/sympy/sympy/wiki/Writing-Release-Notes for more information
on how to write release notes. The bot will check your release notes
automatically to see if they are formatted correctly. -->

<!-- BEGIN RELEASE NOTES -->
- parsing
    - Added support for more data types in C parser: `signed char`, `unsigned char`, `short`, `short int`, `signed short`, `signed short int`, `unsigned short`, `unsigned short int`, `unsigned int`, `long`, `long int`, `signed long`, `signed long int`, `unsigned long`, `unsigned long int`, `double`, `long double`. Also, modified the data type for `int` and `float` to `intc` and `float32` respectively, replacing `integer` and `real`
    - Added support for shorthand operators in C parser: `+=`, `-=`, `*=`, `/=` and `%=`
    - Removed the assumption of value of a variable, if it is not initialized while variable declaration(e.g.; in case of `int a;`, the value of `a` was assumed to be `Integer(0)`, in case of `float b;`, the value of `b` was assumed to be `Float(0.0)` and in case of `bool c;`, the value of `c` was assumed to be `S.false`). Also, removed the assumption of default value of formal parameters while parameter declaration in function definition or function prototype (e.g.; in case of function definition `void func(int a, float b) { //some code }`, default value of formal parameters`a` and `b` were assumed to be `Integer(0)` and `Float(0.0)` respectively)
 
<!-- END RELEASE NOTES -->

smitgajjar committed:
parsing(c_parser): added support for major data types and shorthand op

smitgajjar committed:
Removed unused imports and variables

smitgajjar committed:
Added more tests for data types

smitgajjar commented:
Ping @Sc0rpi0n101 @certik

certik commented:
@Sc0rpi0n101 let's try to finish this soon, so that we can merge it. Please ping me once you are done with reviews.

Sc0rpi0n101 commented on sympy/parsing/c/c_parser.py:
```diff
@@ -234,16 +236,43 @@ def transform_var_decl(self, node):
+                        elif node.type.kind == cin.TypeKind.SHORT:
+                            type = int16
+                        elif node.type.kind == cin.TypeKind.INT:
+                            type = IntBaseType(String('integer'))
```
you can use `intc` for this

smitgajjar replied:
Yes, that would be better as it replicates `int` of C, working on it!

Sc0rpi0n101 commented on sympy/parsing/c/c_parser.py:
```diff
@@ -234,16 +236,43 @@ def transform_var_decl(self, node):
+                        elif node.type.kind == cin.TypeKind.USHORT:
+                            type = uint16
+                        elif node.type.kind == cin.TypeKind.UINT:
+                            type = uint32
```
I'm not sure if this is better or `intp`, although I'd lean towards `uint32` as it contains size information

smitgajjar replied:
Yeah...

Sc0rpi0n101 commented on sympy/parsing/c/c_parser.py:
```diff
@@ -261,8 +290,14 @@ def transform_var_decl(self, node):
-                        type = FloatBaseType(String('real'))
+                    elif node.type.kind in supported_float_types:
+                        if node.type.kind == cin.TypeKind.FLOAT:
+                            type = FloatBaseType(String('real'))
```
you can import `real` from codegen instead of declaring it here again

smitgajjar replied:
Yeah, right :sweat_smile:

smitgajjar replied:
But, I am confused among `real` and `float32`, `intc` and `integer`. Which should I choose? According to the comments, for `real` and `integer`, precision will be chosen by the code printers!

smitgajjar replied:
Once we decide which one to use in these cases(whether `real` or `float32` in case of `float` and `integer` or `intc` or `int32` in case of `int`), the other code(e.g.; `parm_decl` and `func_decl`), which has `IntBaseType` and `FloatBaseType`, also will have to be updated I guess

Sc0rpi0n101 replied:
We can use `intc` and `float32` for now then, and you can update the other visitors accordingly if it's required.

smitgajjar replied:
Hi, I found a shortcoming of `parm_decl` while updating the data types there. It is not possible to get the default value of parameters from python bindings!

Sc0rpi0n101 commented on sympy/parsing/c/c_parser.py:
```diff
@@ -234,16 +236,43 @@ def transform_var_decl(self, node):
 
-                while child.kind == cin.CursorKind.TYPE_REF:
-                    child = next(children)
+                # while child.kind == cin.CursorKind.TYPE_REF:
+                #     child = next(children)
```
any reason for commenting this out?

Sc0rpi0n101 replied:
I'd suggest to leave it as it is. We don't need type references for primary data types anyway.

smitgajjar replied:
Alright... Actually, I was trying to parse the types from `stdint.h` such as `int8_t`, `int16_t` which are exactly similar to `int8`, `int16` of `numpy`. I left that as comment accidentally.

Sc0rpi0n101 reviewed (commented):
How about using dictionaries to manage data types, eg: 
```python
int_types = {cin.TypeKind.SCHAR:int8,
                    cin.TypeKind.SHORT:int16, ...}
if node.type.kind in int_types:
    type = int_types[node.type.kind]
```

smitgajjar committed:
Updated types in var_decl, function_decl and parm_decl
- Defined a dictionary _data_types as an an instance variable
- Defined a dictionary ret_types in function_decl
- Changed integer to intc across all visitor functions
- Changed real to float32 across all visitor functions

smitgajjar committed:
added some more types in ret_types

smitgajjar committed:
updated tests according to the previous two commits

Sc0rpi0n101 commented on sympy/parsing/c/c_parser.py:
```diff
@@ -43,11 +43,11 @@
-        FunctionDefinition, FunctionCall, none, Return, Assignment,
-        Type, int8, int16, int64, uint8, uint16, uint32, uint64,
-        float64, float80, aug_assign)
+    from sympy.codegen.ast import (Variable, Integer, Float,
```
is String` not used anywhere anymore?

smitgajjar replied:
No, all data types have been replaced!

Sc0rpi0n101 commented on sympy/parsing/c/c_parser.py:
```diff
@@ -365,18 +356,18 @@ def transform_var_decl(self, node):
-                    type = IntBaseType(String('integer'))
+                if node.type.kind in self._data_types["int"]:
+                    type = self._data_types["int"][node.type.kind]
                     value = Integer(0)
```
Can you just remove the values here and from the return statement also?
I think it'll better to just return a Variable with its type without a default value when it was not given a default value initially.

smitgajjar replied:
Ok

Sc0rpi0n101 commented on sympy/parsing/c/c_parser.py:
```diff
@@ -400,16 +391,46 @@ def transform_function_decl(self, node):
-                ret_type = IntBaseType(String('integer'))
-            elif (c_ret_type == 'float'):
-                ret_type = FloatBaseType(String('real'))
+            tokens_spelling = [
+            token.spelling for token in node.get_tokens()]
+
+            curr_ret_type = ''
+            for spelling in tokens_spelling:
+                if spelling == node.spelling:
+                    curr_ret_type = curr_ret_type[:-1]
+                    break
+                curr_ret_type += spelling + ' '
```
I just checked, you can use `node.result_type` to get the return type for the function and compare `node.result_type.kind` with self._data_types`. There's no need for a new dict either

smitgajjar replied:
Woah... that's great indeed! No need of that dict now...

Sc0rpi0n101 commented on sympy/parsing/c/c_parser.py:
```diff
@@ -476,12 +497,18 @@ def transform_parm_decl(self, node):
-                type = IntBaseType(String('integer'))
+            if node.type.kind in self._data_types["int"]:
+                type = self._data_types["int"][node.type.kind]
                 value = Integer(0)
```
Can you remove the values from here too.

Sc0rpi0n101 replied:
Also remove it from the param definition at line 548.

smitgajjar replied:
Ok

Sc0rpi0n101 commented on sympy/parsing/c/c_parser.py:
```diff
@@ -494,10 +521,18 @@ def transform_parm_decl(self, node):
                 lit = self.transform(child)
-                if (node.type.kind == cin.TypeKind.INT):
+                if node.type.kind in self._data_types["int"]:
+                    type = self._data_types["int"][node.type.kind]
```
and type from here as they are already defined above

smitgajjar replied:
That seems precise!

Sc0rpi0n101 reviewed (commented):
You need to update the doctests in SymPyExpression too, that's why the tests are failing.

Sc0rpi0n101 commented on sympy/parsing/c/c_parser.py:
```diff
@@ -365,16 +391,46 @@ def transform_function_decl(self, node):
-                ret_type = IntBaseType(String('integer'))
-            elif (c_ret_type == 'float'):
-                ret_type = FloatBaseType(String('real'))
+            tokens_spelling = [
+            token.spelling for token in node.get_tokens()]
```
Also, do these definitions really need to be in two lines? this is just increasing the size of the file.

smitgajjar replied:
I thought to follow PEP8 strictly, that line exceeded 72 characters :sweat_smile:

smitgajjar committed:
Removed redundancies, improved current impl and updated tests
- Removed default value of int (was 0), float (was 0.0) and bool(was false) when the variables are declared, but not initialized and also in case of assumption of values of default parameters
- Removed type of parameter once, which was assigned twice in the param_decl
- Removed dictionary ret_types, instead used node.result_type.kind and self._data_types
- Updated tests for above changes

smitgajjar committed:
parsing(sym_expr.py): Updated doctest eg, not to initialize integer to 0

smitgajjar committed:
parsing(test_sym_expr): Updated test, not to initialize default val

smitgajjar committed:
removed redundant variable value

Sc0rpi0n101 commented on sympy/parsing/c/c_parser.py:
```diff
@@ -308,12 +337,12 @@ def transform_var_decl(self, node):
-                        type = FloatBaseType(String('real'))
-                    elif (node.type.kind == cin.TypeKind.BOOL):
-                        type = Type(String('bool'))
+                    if node.type.kind in self._data_types["int"]:
+                        type = self._data_types["int"][node.type.kind]
+                    elif node.type.kind in self._data_types["float"]:
+                        type = self._data_types["float"][node.type.kind]
+                    elif node.type.kind in self._data_types["bool"]:
+                        type = self._data_types["bool"][node.type.kind]
```
I feel it might be better to move all the type declarations to before the try statement, It'll help reduce a lot of the redundancy.
Just keep the value declarations here

smitgajjar replied:
Looks like, it will indeed reduce redundanies, working on it now!

Sc0rpi0n101 commented on sympy/parsing/c/c_parser.py:
```diff
@@ -330,25 +359,30 @@ def transform_var_decl(self, node):
-                type = type,
-                value = value
-            )
+            if value == None:
+                return Variable(
+                    node.spelling
+                ).as_Declaration(
+                    type = type
+                )
+            else:
+                return Variable(
+                    node.spelling
+                ).as_Declaration(
+                    type = type,
+                    value = value
+                )
```
Both of these return the same thing. Value defaults to NoneToken anyway when not provided explicitly.

smitgajjar replied:
Ohh, great one!

Sc0rpi0n101 commented on sympy/parsing/c/c_parser.py:
```diff
@@ -330,25 +359,30 @@ def transform_var_decl(self, node):
                     raise NotImplementedError("Only bool, int "
-                            "and float are supported")
+                        "and float are supported")
+                value = None
```
no need of that either, just remove value completely from here

smitgajjar replied:
Alright then

smitgajjar replied:
Does that look good? I didn't remove it though. :sweat_smile: See whether it is better or not

Sc0rpi0n101 replied:
This portion of the code is executed when the variables are declared without a default value like `int a` or `float b`
as the var_decl node will have no children in that case. So it'll be better to just declare the Variable without a 'value'. I think it'll be better to return without a value.

smitgajjar replied:
I got it, I thought, when there is only declaration, it will be ok to have `None`.  I am changing it now, let me know, if that looks good

Sc0rpi0n101 replied:
It is okay to have None, but why waste a new variable and a few extra lines of code when it's going to default to that anyway.

smitgajjar committed:
Removed redundancies in var_decl
- Moved "type" variables out of try block to remove repeated assignments in every if statement
- Removed removed if condition to return Variable in var_decl

Sc0rpi0n101 commented on sympy/parsing/c/c_parser.py:
```diff
@@ -270,8 +279,6 @@ def transform_var_decl(self, node):
-
                         # when only one decl_ref_expr is assigned
                         # e.g., int b = a;
                         if isinstance(val, str):
```
There is still a bit of redundancy in the code.

I think it might be better to make these the outer if statements as `value` is the same in case of strings and the else statements. The other if statements can have the `if node.type.kind in self._data_types["int"]` as nested statements accordingly.

smitgajjar replied:
Looking into it, how those redundancies can be removed

smitgajjar replied:
Hey, don't you think, in that case, we will have to write `if node.type.kind in self._data_types["blah"]:` again inside those two if statements, resulting in a redundancy again

smitgajjar replied:
See:
``` python
                    if isinstance(val, str):
                        value = Symbol(val)
                    elif isinstance(val, bool):
                        if node.type.kind in self._data_types["int"]:
                            value = Integer(0) if val == False else Integer(1)
                        elif node.type.kind in self._data_types["float"]:
                            value = Float(0.0) if val == False else Float(1.0)
                        elif node.type.kind in self._data_types["bool"]:
                            value = sympify(val)
                    elif isinstance(val, (Integer, int, Float, float)):
                        if node.type.kind in self._data_types["int"]:
                            value = Integer(val)
                        elif node.type.kind in self._data_types["float"]:
                            value = Float(val)
                        elif node.type.kind in self._data_types["bool"]:
                            value = sympify(bool(val))
                    else:
                        value = val
```

Sc0rpi0n101 replied:
Itis also less code than previous and nothing is being repeated at least. The if statements are repeated but they have different statements

smitgajjar committed:
Refactored return part of var_decl

Sc0rpi0n101 commented on sympy/parsing/c/c_parser.py:
```diff
@@ -329,26 +349,18 @@ def transform_var_decl(self, node):
                 node.spelling
-            ).as_Declaration(
-                type = type,
-                value = value
-            )
+                ).as_Declaration(
+                    type = type,
+                    value = value
+                )
```
You can even have this return statement inside the try block, it might look a little more organized. But it doesn't make much difference tbh. You can leave it here too if you want to.

smitgajjar replied:
Yeah :)

smitgajjar committed:
Refactored and improved transform_var_decl

smitgajjar committed:
Maintained regular tabs in return statement

Sc0rpi0n101 commented on sympy/parsing/tests/test_c_parser.py:
```diff
@@ -3038,18 +3550,10 @@ def test_var_decl():
-        )
-
-        c_src_raise3 = (
             'int a = (int) 1.0;'
         )
 
-        c_src_raise4 = (
+        c_src_raise2 = (
             'int a = 10;'
             'int b = (a!=10)?(a):(0);'
         )
```
These are giving error outputs on the tests, can you just remove these too,they're not being tested anyway? or maybe try some other tests like data types that have not been implemented yet, but the node visitors like transform_var_decl have been implemented?

smitgajjar replied:
Yeah, it raises these errors
<pre>
====================================================================================== test process starts =======================================================================================
executable:         /home/smit/anaconda3/bin/python  (3.7.1-final-0) [CPython]
architecture:       64-bit
cache:              yes
ground types:       gmpy 2.0.8
numpy:              1.15.4
random seed:        87512398
hash randomization: on (PYTHONHASHSEED=3615327077)

sympy/parsing/tests/test_c_parser.py[13] .........Ignoring node of type CursorKind.CSTYLE_CAST_EXPR (( int ) 1.0)
Ignoring node of type CursorKind.CONDITIONAL_OPERATOR (( a != 10 ) ? ( a ) : ( 0 ))
....
</pre>

These nodes are unrecognizable, that's why! Alright I would remove these in the next commit...

Sc0rpi0n101 commented on sympy/parsing/c/c_parser.py:
```diff
@@ -227,6 +252,15 @@ def transform_var_decl(self, node):
             types
```
Can you also update the docstring with the supported data types?

smitgajjar replied:
Sure... In the next commit!

smitgajjar committed:
Added more supported data types in docstring of var_decl

smitgajjar committed:
Updated tests to handle exceptions in var_decl

Sc0rpi0n101 reviewed (approved):
LGTM
ping @certik

smitgajjar commented:
Ping!

Sc0rpi0n101 commented:
ping again @certik

certik commented:
I think this looks good, thank you @smitgajjar for the PR and thank you @Sc0rpi0n101 for reviewing!
